src/magic_folder/scanner.py

```diff
@@ -0,0 +1,129 @@
+from .util.file import get_pathinfo
+from .util.twisted import exclusively
+
+
+def _create_cooperator(clock):
+    """
+    Create a cooperator
+    """
+
+    def schedule(f):
+        return clock.callLater(0, f)
+
+    # NOTE: We don't use CooperatorSevice here, since:
+    # - There is not a way to set the reactor it uses
+    # - Once we enable periodic scans, we want to wait to
+    #   stop it until after it has We want to stop it after the TimerService has
+    #   stoppped, so that it will have no pending work.
+    return Cooperator(
+        scheduler=schedule,
+    )
+
+
+@attr.s
+class ScannerService(Service):
+    """
+    Periodically scan a local Magic Folder for new or updated files
+    """
+
+    _config = attr.ib()
+    _local_snapshot_service = attr.ib()
+    _status = attr.ib()
+    _cooperator = attr.ib()
+    _lock = attr.ib(init=False, factory=DeferredLock)
+
+    @classmethod
+    def from_config(cls, clock, folder_config, local_snapshot_service, status):
+        return cls(
+            config=folder_config,
+            local_snapshot_service=local_snapshot_service,
+            status=status,
+            cooperator=_create_cooperator(clock),
+        )
+
+    def __attrs_post_init__(self):
+        super(ScannerService, self).__init__()
+
+    def stopService(self):
+        super(ScannerService, self).stopService()
+        self._cooperator.stop()
+
+    def scan_once(self):
+        """
+        Perform a scan for new files.
+        """
+        return self._scan()
+
+    @exclusively
+    @inline_callbacks
+    def _scan(self):
+        """
+        Perform a scan for new files, and wait for all the snapshots to be
+        complete.
+        """
+        # TODO: Do we always want to wait for all the files to be snapshotted?
+        # If we don't wait for snapshotting to complete, then we probably want
+        # to coalesce multiple outstanding snapshot requests for the same file,
+        # or otherwise ensure that we don't end up snapshotting an unchanged
+        # file.
+        results = []
+
+        def process(path):
+            d = self._local_snapshot_service.add_file(path)
+            d.addErrback(write_failure)
+            results.append(d)
+
+        with start_action(action_type="scanner:find-updates"):
+            yield find_updated_files(self._cooperator, self._config, process)
+            yield gatherResults(results)
+        # XXX update/use IStatus to report scan start/end
+
+
+def find_updated_files(cooperator, folder_config, on_new_file):
+    """
+    :param Cooperator cooperator: The cooperator to use to control yielding to
+        the reactor.
+    :param MagicFolderConfig folder_config: the folder for which we
+        are scanning
+
+    :param callable on_new_file: a 1-argument callable. This function
+        will be invoked for each updated / new file we find. The
+        argument will be a FilePath of the updated/new file.
+
+    :returns: the scan duration, in seconds
```

Comment:
I think it returns a `Deferred` that fires with ... `TaskDone()` maybe?  Not sure exactly what happens with the version of Twisted we're using, actually.  Probably not a scan duration in seconds though.